security/crypto/bls12_test.go

```diff
@@ -0,0 +1,219 @@
+package crypto_test
+
+import (
+	"testing"
+
+	"github.com/relab/hotstuff"
+	"github.com/relab/hotstuff/core"
+	"github.com/relab/hotstuff/internal/testutil"
+	"github.com/relab/hotstuff/security/crypto"
+)
+
+// createBLS12Signers creates BLS12-based signers for testing.
+// Returns a slice of signers, one for each replica.
+func createBLS12Signers(t testing.TB, numReplicas int) []crypto.Base {
+	t.Helper()
+	// Generate keys for all replicas
+	keys := make(map[hotstuff.ID]hotstuff.PrivateKey, numReplicas)
+	for i := 1; i <= numReplicas; i++ {
+		keys[hotstuff.ID(i)] = testutil.GenerateBLS12Key(t)
+	}
+	
+	// Create configs for all replicas
+	configs := make([]*core.RuntimeConfig, numReplicas)
+	for i := 1; i <= numReplicas; i++ {
+		id := hotstuff.ID(i)
+		configs[i-1] = core.NewRuntimeConfig(id, keys[id])
+		for j := 1; j <= numReplicas; j++ {
+			rid := hotstuff.ID(j)
+			configs[i-1].AddReplica(&hotstuff.ReplicaInfo{ID: rid, PubKey: keys[rid].Public()})
+		}
+	}
+	
+	// Create base crypto instances (this adds connection metadata for each)
+	bases := make([]crypto.Base, numReplicas)
+	for i := 1; i <= numReplicas; i++ {
+		base, err := crypto.NewBLS12(configs[i-1])
+		if err != nil {
+			t.Fatalf("Failed to create BLS12 base: %v", err)
+		}
+		bases[i-1] = base
+	}
+	
+	// Share proof-of-possession metadata between all replicas
+	for i := 1; i <= numReplicas; i++ {
```

Comment:
Since `i-1` and `j-1` is used to index into the slices, we should change these to:
```go
for i := range numReplicas {
```
And
```go
for j := range numReplicas {
```
And then update to `hotstuff.ID(i+1)` and similar for the `j`.